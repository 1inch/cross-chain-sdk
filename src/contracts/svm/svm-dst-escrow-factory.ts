import {BN, BorshCoder, web3} from '@coral-xyz/anchor'
import assert from 'assert'
import {Buffer} from 'buffer'
import {Instruction} from './instruction'
import {BaseProgram} from './base-program'
import {WhitelistContract} from './whitelist'
import {ParsedCreateDstEscrowInstructionData} from './types'
import {bigintToBN} from '../../utils/numbers/bigint-to-bn'
import {uintAsBeBytes} from '../../utils/numbers/uint-as-be-bytes'
import {uint256split} from '../../utils/numbers/uint256-split'
import {getAta, getPda} from '../../utils'
import {
    EvmAddress,
    HashLock,
    Immutables,
    SolanaAddress,
    TimeLocks
} from '../../domains'
import {IDL} from '../../idl/cross-chain-escrow-dst'
import {FixedLengthArray} from '../../type-utils'
import {bufferToHex} from '../../utils/bytes'
import {bnArrayToBigInt} from '../../utils/numbers/bn-array-to-big-int'

export class SvmDstEscrowFactory extends BaseProgram {
    static DEFAULT = new SvmDstEscrowFactory(
        new SolanaAddress('GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve')
    )

    private static readonly coder = new BorshCoder(IDL)

    constructor(programId: SolanaAddress) {
        super(programId)
    }

    static parseCreateEscrowInstruction(
        ix: Instruction
    ): ParsedCreateDstEscrowInstructionData {
        const decoded = SvmDstEscrowFactory.coder.instruction.decode(
            ix.data
        ) as {
            name: string
            data: {
                orderHash: FixedLengthArray<number, 32>
                hashlock: FixedLengthArray<number, 32>
                amount: BN
                safetyDeposit: BN
                recipient: web3.PublicKey
                timelocks: FixedLengthArray<BN, 4>
                srcCancellationTimestamp: number
                assetIsNative: boolean
            }
        }

        assert(decoded, 'cannot decode create instruction')
        assert(decoded.name === 'create', 'not create instruction')

        const data = decoded.data

        return {
            orderHash: bufferToHex(data.orderHash),
            hashlock: HashLock.fromString(bufferToHex(data.hashlock)),
            amount: BigInt(data.amount.toString()),
            safetyDeposit: BigInt(data.safetyDeposit.toString()),
            recipient: EvmAddress.fromBuffer(data.recipient.toBuffer()),
            timelocks: TimeLocks.fromBigInt(bnArrayToBigInt(data.timelocks)),
            srcCancellationTimestamp: data.srcCancellationTimestamp,
            assetIsNative: data.assetIsNative
        }
    }

    static parsePrivateWithdrawInstruction(ix: Instruction): {secret: string} {
        const decoded = this.coder.instruction.decode(ix.data) as {
            name: string
            data: {secret: FixedLengthArray<number, 32>}
        }

        assert(decoded, 'cannot decode withdraw instruction')
        assert(decoded.name === 'withdraw', 'not withdraw instruction')

        return {
            secret: bufferToHex(decoded.data.secret)
        }
    }

    static parsePublicWithdrawInstruction(ix: Instruction): {secret: string} {
        const decoded = this.coder.instruction.decode(ix.data) as {
            name: string
            data: {secret: FixedLengthArray<number, 32>}
        }

        assert(decoded, 'cannot decode publicWithdraw instruction')
        assert(
            decoded.name === 'publicWithdraw',
            'not publicWithdraw instruction'
        )

        return {
            secret: bufferToHex(decoded.data.secret)
        }
    }

    public getEscrowAddress(params: EscrowAddressParams): SolanaAddress {
        return getPda(this.programId, [
            this.encoder.encode('escrow'),
            params.orderHash,
            params.hashLock.toBuffer(),
            params.maker.toBuffer(),
            uintAsBeBytes(params.amount, 64)
        ])
    }

    public createEscrow(
        params: Immutables<SolanaAddress>,
        extra: {
            /**
             * TokenProgram or TokenProgram 2022
             */
            tokenProgramId: SolanaAddress
            srcCancellationTimestamp: bigint
        }
    ): Instruction {
        const token = params.token.isNative()
            ? SolanaAddress.WRAPPED_NATIVE
            : params.token

        const data = SvmDstEscrowFactory.coder.instruction.encode('create', {
            orderHash: params.orderHash,
            hashlock: params.hashLock.toBuffer(),
            amount: new BN(params.amount.toString()),
            safetyDeposit: new BN(params.safetyDeposit.toString()),
            recipient: params.maker,
            timelocks: uint256split(params.timeLocks.build()).map(bigintToBN),
            srcCancellationTimestamp: Number(extra.srcCancellationTimestamp),
            assetIsNative: params.token.isNative()
        })

        const escrowAddress = this.getEscrowAddress(params)

        return new Instruction(
            this.programId,
            [
                {
                    // 1. taker
                    pubkey: params.taker,
                    isSigner: true,
                    isWritable: true
                },
                {
                    // 2. mint
                    pubkey: token,
                    isWritable: false,
                    isSigner: false
                },
                this.optionalAccount(
                    {
                        // 3. taker_ata
                        pubkey: getAta(
                            params.taker,
                            token,
                            extra.tokenProgramId
                        ),
                        isSigner: false,
                        isWritable: true
                    },
                    params.token.isNative()
                ),
                {
                    // 4. escrow
                    pubkey: escrowAddress,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 5. escrow ata
                    pubkey: getAta(escrowAddress, token, extra.tokenProgramId),
                    isWritable: true,
                    isSigner: false
                },

                {
                    // 6. associated_token_program
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 7. token_program
                    pubkey: extra.tokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 8. rent
                    pubkey: SolanaAddress.SYSVAR_RENT_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 9. system_program
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }

    public withdrawPrivate(
        params: Immutables<SolanaAddress>,
        secret: Buffer,
        extra: {
            /**
             * TokenProgram or TokenProgram 2022
             */
            tokenProgramId: SolanaAddress
        }
    ): Instruction {
        const token = params.token.isNative()
            ? SolanaAddress.WRAPPED_NATIVE
            : params.token
        const data = SvmDstEscrowFactory.coder.instruction.encode('withdraw', {
            secret
        })
        const escrow = this.getEscrowAddress(params)

        return new Instruction(
            this.programId,
            [
                {
                    // 1. taker
                    pubkey: params.taker,
                    isSigner: true,
                    isWritable: true
                },
                {
                    // 2. recipient
                    pubkey: params.maker,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 3. dst token
                    pubkey: token,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 4. escrow
                    pubkey: escrow,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 5. escrow_ata
                    pubkey: getAta(escrow, token, extra.tokenProgramId),
                    isSigner: false,
                    isWritable: true
                },
                this.optionalAccount(
                    {
                        // 6. recipient_ata
                        pubkey: getAta(
                            params.maker,
                            params.token,
                            extra.tokenProgramId
                        ),
                        isSigner: false,
                        isWritable: true
                    },
                    params.token.isNative()
                ),
                {
                    // 7. associated_token_program
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 8. token_program
                    pubkey: extra.tokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 9. system_program
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }

    public withdrawPublic(
        params: Immutables<SolanaAddress>,
        secret: Buffer,
        payer: SolanaAddress,
        extra: {
            /**
             * If not passed, than `WhitelistContract.DEFAULT` will be used
             * @see WhitelistContract.DEFAULT
             */
            whitelistProgramId?: SolanaAddress
            /**
             * TokenProgram or TokenProgram 2022
             */
            tokenProgramId: SolanaAddress
        }
    ): Instruction {
        const whitelistProgram = extra.whitelistProgramId
            ? new WhitelistContract(extra.whitelistProgramId)
            : WhitelistContract.DEFAULT

        const token = params.token.isNative()
            ? SolanaAddress.WRAPPED_NATIVE
            : params.token
        const data = SvmDstEscrowFactory.coder.instruction.encode(
            'publicWithdraw',
            {secret}
        )
        const escrow = this.getEscrowAddress(params)

        return new Instruction(
            this.programId,
            [
                {
                    // 1. creator
                    pubkey: params.taker,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 2. recipient
                    pubkey: params.maker,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 3. payer
                    pubkey: payer,
                    isSigner: true,
                    isWritable: true
                },
                {
                    // 4. resolver_access
                    pubkey: whitelistProgram.getAccessAccount(payer),
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 5. mint
                    pubkey: token,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 6. escrow
                    pubkey: escrow,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 7. escrow_ata
                    pubkey: getAta(escrow, token, extra.tokenProgramId),
                    isSigner: false,
                    isWritable: true
                },
                this.optionalAccount(
                    {
                        // 8. recipient_ata
                        pubkey: getAta(
                            params.maker,
                            params.token,
                            extra.tokenProgramId
                        ),
                        isSigner: false,
                        isWritable: true
                    },
                    params.token.isNative()
                ),
                {
                    // 9. associated_token_program
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 10. token_program
                    pubkey: extra.tokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 11. system_program
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }
}
export type EscrowAddressParams = Pick<
    Immutables<SolanaAddress>,
    'orderHash' | 'hashLock' | 'maker' | 'amount'
>
