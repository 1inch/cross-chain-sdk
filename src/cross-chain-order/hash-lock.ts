import {id} from 'ethers'
import {getBytesCount} from '@1inch/byte-utils'
import assert from 'assert'

export class HashLock {
    public static Web3Type = 'bytes32'

    private readonly value: string

    protected constructor(val: string) {
        this.value = val
    }

    /**
     * Create HashLock from keccak256 hash of secret
     */
    public static fromSecret(secret: string): HashLock {
        return new HashLock(id(secret))
    }

    public static fromString(hashLock: string): HashLock {
        assert(
            getBytesCount(hashLock) === 32n,
            'HashLock string must be result of keccak256'
        )

        return new HashLock(hashLock)
    }

    public toString(): string {
        return this.value
    }
}
