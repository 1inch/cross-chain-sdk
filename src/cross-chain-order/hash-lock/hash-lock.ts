import {keccak256, solidityPackedKeccak256} from 'ethers'
import {SimpleMerkleTree} from '@openzeppelin/merkle-tree'
import {BitMask, BN, getBytesCount, isHexBytes} from '@1inch/byte-utils'
import assert from 'assert'

export class HashLock {
    public static Web3Type = 'bytes32'

    private readonly value: string

    protected constructor(val: string) {
        this.value = val
    }

    public static hashSecret(secret: string): string {
        assert(
            isHexBytes(secret) && getBytesCount(secret) === 32n,
            'secret length must be 32 bytes hex encoded'
        )

        return keccak256(secret)
    }

    public static getMerkleLeaves(secrets: string[]): MerkleLeaf[] {
        return HashLock.getMerkleLeavesFromSecretHashes(
            secrets.map(HashLock.hashSecret)
        )
    }

    public static getMerkleLeavesFromSecretHashes(
        secretHashes: string[]
    ): MerkleLeaf[] {
        return secretHashes.map(
            (s, idx) =>
                solidityPackedKeccak256(
                    ['uint64', 'bytes32'],
                    [idx, s]
                ) as MerkleLeaf
        )
    }

    public static getProof(leaves: string[], idx: number): MerkleLeaf[] {
        return SimpleMerkleTree.of(leaves).getProof(idx) as MerkleLeaf[]
    }

    public static fromString(value: string): HashLock {
        assert(
            isHexBytes(value) && getBytesCount(value) === 32n,
            'HashLock value must be bytes32 hex encoded'
        )

        return new HashLock(value)
    }

    /**
     * Create HashLock from keccak256 hash of secret
     */
    public static forSingleFill(secret: string): HashLock {
        return new HashLock(HashLock.hashSecret(secret))
    }

    public static forMultipleFills(leaves: MerkleLeaf[]): HashLock {
        assert(
            leaves.length > 2,
            'leaves array must be greater than 2. Or use HashLock.forSingleFill'
        )
        const root = SimpleMerkleTree.of(leaves).root
        const rootWithCount = BN.fromHex(root).setMask(
            new BitMask(240n, 256n),
            BigInt(leaves.length - 1)
        )

        return new HashLock(rootWithCount.toHex(64))
    }

    /**
     * Only use if HashLockInfo is for multiple fill order
     * Otherwise garbage will be returned
     *
     */
    public getPartsCount(): bigint {
        return new BN(BigInt(this.value)).getMask(new BitMask(240n, 256n)).value
    }

    public toString(): string {
        return this.value
    }

    public eq(other: HashLock): boolean {
        return this.value === other.value
    }
}

export type MerkleLeaf = string & {_tag: 'MerkleLeaf'}
