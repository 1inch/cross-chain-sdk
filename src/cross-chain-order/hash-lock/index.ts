export {HashLock} from './hash-lock'
