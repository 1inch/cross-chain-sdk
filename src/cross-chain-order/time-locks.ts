import {BitMask, BN, UINT_32_MAX} from '@1inch/byte-utils'
import assert from 'assert'

// todo check intervals and add docs
export class TimeLocks {
    static Web3Type = 'uint256'

    protected constructor(
        public readonly deployedAt: bigint,
        private readonly _srcWithdrawal: bigint,
        private readonly _srcPublicWithdrawal: bigint,
        private readonly _srcCancellation: bigint,
        private readonly _srcPublicCancellation: bigint,
        private readonly _dstWithdrawal: bigint,
        private readonly _dstPublicWithdrawal: bigint,
        private readonly _dstCancellation: bigint
    ) {
        assert(
            deployedAt <= UINT_32_MAX,
            'deployedAt can not be > uint32 max value'
        )

        assert(
            _srcWithdrawal <= UINT_32_MAX,
            'srcWithdrawal can not be > uint32 max value'
        )
        assert(
            _srcPublicWithdrawal <= UINT_32_MAX,
            'srcPublicWithdrawal can not be > uint32 max value'
        )
        assert(
            _srcCancellation <= UINT_32_MAX,
            'srcCancellation can not be > uint32 max value'
        )
        assert(
            _srcPublicCancellation <= UINT_32_MAX,
            'srcPublicCancellation can not be > uint32 max value'
        )
        assert(
            _dstWithdrawal <= UINT_32_MAX,
            'dstWithdrawal can not be > uint32 max value'
        )
        assert(
            _dstPublicWithdrawal <= UINT_32_MAX,
            'dstPublicWithdrawal can not be > uint32 max value'
        )
        assert(
            _dstCancellation <= UINT_32_MAX,
            'dstCancellation can not be > uint32 max value'
        )
    }

    public get srcWithdrawal(): bigint {
        return this.deployedAt + this._srcWithdrawal
    }

    public get srcPublicWithdrawal(): bigint {
        return this.deployedAt + this._srcPublicWithdrawal
    }

    public get srcCancellation(): bigint {
        return this.deployedAt + this._srcCancellation
    }

    public get srcPublicCancellation(): bigint {
        return this.deployedAt + this._srcPublicCancellation
    }

    public get dstWithdrawal(): bigint {
        return this.deployedAt + this._dstWithdrawal
    }

    public get dstPublicWithdrawal(): bigint {
        return this.deployedAt + this._dstPublicWithdrawal
    }

    public get dstCancellation(): bigint {
        return this.deployedAt + this._dstCancellation
    }

    public static new(params: {
        deployedAt: bigint
        srcWithdrawal: bigint
        srcPublicWithdrawal: bigint
        srcCancellation: bigint
        srcPublicCancellation: bigint
        dstWithdrawal: bigint
        dstPublicWithdrawal: bigint
        dstCancellation: bigint
    }): TimeLocks {
        return new TimeLocks(
            params.deployedAt,
            params.srcWithdrawal,
            params.srcPublicWithdrawal,
            params.srcCancellation,
            params.srcPublicCancellation,
            params.dstWithdrawal,
            params.dstPublicWithdrawal,
            params.dstCancellation
        )
    }

    public static fromBigInt(val: bigint): TimeLocks {
        const valBN = new BN(val)

        const params = Array.from({length: 8}).map((_, i) => {
            return valBN.getMask(
                new BitMask(BigInt(i) * 32n, BigInt(i + 1) * 32n)
            ).value
        }) as [bigint, bigint, bigint, bigint, bigint, bigint, bigint, bigint] // ts cannot infer that we have exactly 8 elements in the array after map

        return new TimeLocks(...params)
    }

    public build(): bigint {
        return [
            this.deployedAt,
            this._srcWithdrawal,
            this._srcPublicWithdrawal,
            this._srcCancellation,
            this._srcPublicCancellation,
            this._dstWithdrawal,
            this._dstPublicWithdrawal,
            this._dstCancellation
        ].reduceRight((acc, el) => (acc << 32n) | el)
    }
}
