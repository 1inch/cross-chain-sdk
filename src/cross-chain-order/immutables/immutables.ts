import {Address} from '@1inch/fusion-sdk'
import {AbiCoder, getCreate2Address, keccak256} from 'ethers'
import {isHexBytes, trim0x} from '@1inch/byte-utils'
import assert from 'assert'
import {HashLock} from '../hash-lock'
import {TimeLocks} from '../time-locks/time-locks'
import {
    ESCROW_DST_IMPLEMENTATION_ADDRESS,
    ESCROW_FACTORY_ADDRESS,
    ESCROW_SRC_IMPLEMENTATION_ADDRESS
} from '../contants'

/**
 * Contract representation of class
 */
export type ImmutablesData = {
    orderHash: string
    hashlock: string
    maker: string
    taker: string
    token: string
    amount: string
    safetyDeposit: string
    timelocks: string
}

/**
 * Contains escrow params for both source and destination chains
 * Determinate addresses of escrow contracts
 */
export class Immutables {
    public static readonly Web3Type = `tuple(${[
        'bytes32 orderHash',
        'bytes32 hashlock',
        'address maker',
        'address taker',
        'address token',
        'uint256 amount',
        'uint256 safetyDeposit',
        'uint256 timelocks'
    ]})`

    private constructor(
        public readonly orderHash: string,
        public readonly hashLock: HashLock,
        public readonly maker: Address,
        public readonly taker: Address,
        public readonly takerAsset: Address,
        public readonly takerAmount: bigint,
        public readonly safetyDeposit: bigint, // todo: separate class
        public readonly timeLocks: TimeLocks
    ) {}

    public static new(params: {
        orderHash: string
        hashLock: HashLock
        maker: Address
        taker: Address
        takerAsset: Address
        takerAmount: bigint
        safetyDeposit: bigint
        timeLocks: TimeLocks
    }): Immutables {
        return new Immutables(
            params.orderHash,
            params.hashLock,
            params.maker,
            params.taker,
            params.takerAsset,
            params.takerAmount,
            params.safetyDeposit,
            params.timeLocks
        )
    }

    /**
     * Create instance from encoded bytes
     * @param bytes 0x prefixed hex string
     */
    public static decode(bytes: string): Immutables {
        assert(isHexBytes(bytes))
        const res = AbiCoder.defaultAbiCoder().decode(
            [Immutables.Web3Type],
            bytes
        )
        const data = res.at(0) as ImmutablesData

        return new Immutables(
            data.orderHash,
            HashLock.fromString(data.hashlock),
            new Address(data.maker),
            new Address(data.taker),
            new Address(data.token),
            BigInt(data.amount),
            BigInt(data.safetyDeposit),
            TimeLocks.fromBigInt(BigInt(data.timelocks))
        )
    }

    /**
     * See https://github.com/1inch/cross-chain-swap/blob/03d99b9604d8f7a5a396720fbe1059f7d94db762/contracts/libraries/ProxyHashLib.sol#L14
     */
    private static calcProxyBytecodeHash(impl: Address): string {
        return keccak256(
            `0x3d602d80600a3d3981f3363d3d373d3d3d363d73${trim0x(impl.toString())}5af43d82803e903d91602b57fd5bf3`
        )
    }

    /**
     * Return keccak256 hash of instance
     */
    public hash(): string {
        return keccak256(this.encode())
    }

    public build(): ImmutablesData {
        return {
            orderHash: this.orderHash,
            hashlock: this.hashLock.toString(),
            maker: this.maker.toString(),
            taker: this.taker.toString(),
            token: this.takerAsset.toString(),
            amount: this.takerAmount.toString(),
            safetyDeposit: this.safetyDeposit.toString(),
            timelocks: this.timeLocks.build().toString()
        }
    }

    /**
     * Encode instance as bytes
     */
    public encode(): string {
        return AbiCoder.defaultAbiCoder().encode(
            [Immutables.Web3Type],
            [this.build()]
        )
    }

    /**
     * Calculate address of source escrow contract (source chain)
     */
    public getSrcEscrowAddress(
        factory = ESCROW_FACTORY_ADDRESS,
        srcImplementation = ESCROW_SRC_IMPLEMENTATION_ADDRESS
    ): Address {
        return new Address(
            getCreate2Address(
                factory.toString(),
                this.hash(),
                Immutables.calcProxyBytecodeHash(srcImplementation)
            )
        )
    }

    /**
     * Calculate address of destination escrow contract (destination chain)
     */
    public getDstEscrowAddress(
        factory = ESCROW_FACTORY_ADDRESS,
        dstImplementation = ESCROW_DST_IMPLEMENTATION_ADDRESS
    ): Address {
        return new Address(
            getCreate2Address(
                factory.toString(),
                this.hash(),
                Immutables.calcProxyBytecodeHash(dstImplementation)
            )
        )
    }
}
