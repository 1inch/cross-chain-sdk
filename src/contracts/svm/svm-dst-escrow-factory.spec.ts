import {parseEther, parseUnits} from 'ethers'
import {SvmDstEscrowFactory} from './svm-dst-escrow-factory'
import {SolanaAddress} from '../../domains/addresses'
import {HashLock} from '../../domains/hash-lock'
import {TimeLocks} from '../../domains/time-locks'
import {Immutables} from '../../domains/immutables'

describe('SVM Escrow dst factory', () => {
    it('should generate withdrawPublic instruction', () => {
        const immutables = Immutables.new({
            orderHash: Buffer.from(
                '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
                'hex'
            ),
            hashLock: HashLock.forSingleFill(
                '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
            ),
            taker: SolanaAddress.fromBigInt(100n),
            token: SolanaAddress.NATIVE, // Native SOL to test wrapped native conversion
            maker: SolanaAddress.fromBigInt(300n),
            amount: parseEther('1'),
            safetyDeposit: 1000n,
            timeLocks: TimeLocks.fromDurations({
                srcFinalityLock: 10n,
                srcPrivateWithdrawal: 200n,
                srcPublicWithdrawal: 100n,
                srcPrivateCancellation: 100n,
                dstFinalityLock: 10n,
                dstPrivateWithdrawal: 100n,
                dstPublicWithdrawal: 100n
            })
        })

        const secret = Buffer.from(
            '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
            'hex'
        )
        const payer = SolanaAddress.fromBigInt(400n)

        const ix = SvmDstEscrowFactory.DEFAULT.withdrawPublic(
            immutables,
            secret,
            payer,
            {
                tokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
            }
        )

        expect(ix).toMatchSnapshot()
    })

    it('should handle native token correctly in withdrawPublic', () => {
        const immutables = Immutables.new({
            orderHash: Buffer.from(
                'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
                'hex'
            ),
            hashLock: HashLock.forSingleFill(
                '0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
            ),
            taker: SolanaAddress.fromBigInt(7777n),
            token: SolanaAddress.NATIVE, // Native SOL
            maker: SolanaAddress.fromBigInt(8888n),
            amount: parseUnits('0.1', 9), // 0.1 SOL
            safetyDeposit: 500n,
            timeLocks: TimeLocks.fromDurations({
                srcFinalityLock: 5n,
                srcPrivateWithdrawal: 100n,
                srcPublicWithdrawal: 80n,
                srcPrivateCancellation: 60n,
                dstFinalityLock: 8n,
                dstPrivateWithdrawal: 90n,
                dstPublicWithdrawal: 70n
            })
        })

        const secret = Buffer.from(
            'bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb',
            'hex'
        )
        const payer = SolanaAddress.fromBigInt(6666n)

        const ix = SvmDstEscrowFactory.DEFAULT.withdrawPublic(
            immutables,
            secret,
            payer,
            {
                tokenProgramId: SolanaAddress.TOKEN_PROGRAM_ID
            }
        )

        expect(ix).toMatchSnapshot()
    })
})
