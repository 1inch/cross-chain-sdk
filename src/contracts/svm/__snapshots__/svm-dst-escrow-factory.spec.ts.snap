// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SVM Escrow dst factory should generate cancelPrivate instruction 1`] = `
{
  "accounts": [
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "11111111111111111111111111111112j",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "7on3yxZPgQgehEo4dxTNg2vV3ECWF8T8QVo3JJSGdcHk",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "FscNvqUJyRy5xqiBPEmC7CnJ6Yw5UaYxLoXmPJgxYw3q",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "2DDSCFzVdQgu5aFSSksZCfD9CvhXFyFNwxMpeFzzdWjg",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "A4zgefDchm9xaifwaEUSrNnB3B96XSNeLJh8s9ACR87o",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": "0xe8dbdf29dbecdcbe",
  "programId": "GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve",
}
`;

exports[`SVM Escrow dst factory should generate create escrow instruction 1`] = `
{
  "accounts": [
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "CTF3k9CXyCjtCvpPMtG24Bo8r8qaQudTgqaSsp6A64tx",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "7on3yxZPgQgehEo4dxTNg2vV3ECWF8T8QVo3JJSGdcHk",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "6EdXXGGUP1nb4dzFkaTXXMAreSSDY18USHGBqT5raivs",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "g2m1CJUtiEZWZsQT6czJCDeJsmAAxc2Rskhdey8wGxP",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "JBx63bJv329f2DkoXNYuvGzCHsvCTKY3MFNiqYHF52pG",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "SysvarRent111111111111111111111111111111111",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": "0x181ec828051c0777663e58c85e0ac6721b6fe635eda03279e9552866f45938dc00effaac1fd0ac0b6e6b426ba6241e7544d5b6802897f93d68e6ed09cbc816058cf67096ff1494de00ca9a3b00000000e803000000000000b96a60f72dcd59fa0e8378cd0f86bbda0a6dc305d1a228705c9d620ada73b62c0a000000d2000000360100009a0100000a0000006e000000d2000000ddbe7b686400000000",
  "programId": "GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve",
}
`;

exports[`SVM Escrow dst factory should generate withdrawPublic instruction 1`] = `
{
  "accounts": [
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "11111111111111111111111111111112j",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "1111111111111111111111111111116B",
    },
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "1111111111111111111111111111117u",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "Aikth6gKPGZbWbgkZSZumpXpBjTQLq2UUMccZ2CRhtSQ",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "So11111111111111111111111111111111111111112",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "ENbPaTRr3J1CNuo877HxbM8bTGZRen19qNhSmpYYeoew",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "HEebJQkKDBXbGwRXPgN5aknqJCWYdG6s3H6Z1wHWz3it",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": "0x9839f0c05223960b0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
  "programId": "GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve",
}
`;

exports[`SVM Escrow dst factory should handle native token correctly in withdrawPublic 1`] = `
{
  "accounts": [
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "1111111111111111111111111111113K6",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "1111111111111111111111111111113eF",
    },
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "1111111111111111111111111111112yw",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "2ZWTkvbTmmyz9KWtYuAkcwHEgp4VXbaRWaYC2atJzVYf",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "So11111111111111111111111111111111111111112",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "EZkxgAXSduM7q3W6J2XxPjg3tgBCZNhWKBQs3xWaEmkB",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "Dkp9yvxVnPoDSLy7q1VCfnEHbKcm79sRCWgaH4t1uCwd",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": "0x9839f0c05223960bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
  "programId": "GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve",
}
`;
